File: 1.Arrays & Hashing/1.Contains Duplicate/1.Contains Duplicate.py
#!/usr/bin/env python3


def containsDuplicate(nums: list[int]) -> bool:
    hashSet = set()
    for n in nums:
        if n in hashSet:
            return True
        hashSet.add(n)
    return False


# By creating a set, a container with no duplicate elements,
# we can add all the elements to the set while checking if there
# are duplicates in O(1) time.

# Overall Time Complexity : O(n)
# Overall Space Complexity: O(n)

File: 1.Arrays & Hashing/1.Contains Duplicate/1.Contains Duplicate.cpp
#include <iostream>
#include <set>
#include <vector>
#include <unordered_set>
using namespace std;

bool containsDuplicate(vector<int>& nums){
//	set<int> hashset;
//	for(vector<int>::const_iterator i = nums.begin(); i != nums.end(); i++)
//		hashset.insert(*i);
//	
//	return hashset.size() != nums.size();
	
	unordered_set<int> s(nums.begin(), nums.end());
	return (s.size() != nums.size());
}

int main(int argc, char *argv[]) {
	vector<int> v = {1, 2, 3, 4, 5, 4};
	cout << containsDuplicate(v);
	
	return 0;
}